### PR TITLE
upgrade: use topological sort to upgrade formulae

### PR DESCRIPTION
**Pull request by FnControlOption**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

#10701

### CODE REVIEW

File: Library/Homebrew/upgrade.rb [hunk 1]
```diff
@@ -42,6 +43,14 @@ def upgrade_formulae(
         end
       end
 
+      dependency_graph = Utils::TopologicalHash.graph_package_dependencies(formulae_to_install)
+      begin
+        formulae_to_install = dependency_graph.tsort & formulae_to_install
+      rescue TSort::Cyclic
+        # Failed to sort formulae topologically because there are cyclic
+        # dependencies. Let FormulaInstaller handle it.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/upgrade.rb:51]: Would be good to at least print this for `homebrew_developer?` and/or use `odebug` here. Do we need to handle this in any other way or will it work even if this fails?

**Review by MikeMcQuaid — COMMENTED** [review]
This is looking great so far! Can you describe more what this is doing differently and why?

File: Library/Homebrew/upgrade.rb (hunk 1 above)
Comment (FnControlOption) [reply Library/Homebrew/upgrade.rb:51]: For developers, `FormulaInstaller` actually raises an error if there are cyclic dependencies, so we technically could raise an error here to save time.

For non-developers, I guess formulae with cyclic dependencies only become a problem when they aren't bottled.

**Comment by FnControlOption** [comment]
Main differences with #10701

- Moved `Utils::Install.graph_dependencies` to `Utils::TopologicalHash.graph_package_dependencies`
- The old `graph_dependencies` only accepted individual formulae or casks, but now it can also accept an array of packages.
- In the original PR, the list of outdated packages is printed from `Upgrade.upgrade_formulae`, but this change was reverted.

File: Library/Homebrew/upgrade.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/upgrade.rb:51]: Yeh, would be good to error out in the same way as `FormulaInstaller` does it here 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]
Thanks again @FnControlOption!

**Pull request merged**